i2c-flash.c

```diff
@@ -36,18 +44,20 @@ ISR(TWI_vect) {
 	}
 
 	// Acknowledge interrupt to let hardware continue.
-	TWCR |= _BV(TWINT);
+	TWCR = _BV(TWEN) | _BV(TWIE) | _BV(TWINT) | _BV(TWEA);
 }
 
 int main() {
 	// Enable internal pull-ups on I2C pins.
-	PORTC = _BV(4) | _BV(5);
+	PORTC = _BV(PORTC4) | _BV(PORTC5);
 
 	// Configure slave address.
 	TWAR = emulated_addr << 1;
 
 	// Turn on TWI in slave mode, enable interrupt.
 	sei();
+	TWSR = TW_NO_INFO & 1;
+	TWBR = 18;
```

Comment:
Neither of these lines should be needed, since they're both for clock setup but we never generate an I2C clock in slave mode. Did they make a difference for you?

Reply:
> Neither of these lines should be needed, since they're both for clock setup but we don't generate an I2C clock in slave mode. Did they make a difference for you?

That might be true if you're using a real Arduino, but I'm using an Adafruit Boarduino, which doesn't have a crystal and is therefore more prone to drift without this.

Reply:
It really shouldn't matter regardless. I2C timing is driven entirely by pulses on the SCL line, which is driven by the master. As a slave device, the AVR isn't timing anything at all. See 22.5.2 in the datasheet on the Bit Rate Generator Unit, which is what both these writes configure:

> This unit controls the period of SCL when operating in a Master mode. … Slave operation does not depend on Bit Rate or Prescaler settings, but the CPU clock frequency in the Slave must be at least 16 times higher than the SCL frequency.

The only thing these lines change are the "Bit Rate or Prescaler settings".

Reply:
IDK, I tried it both ways and its behavior just seemed more stable this way. It could be mere coincidence, but it can't hurt anything.